feat: add Service Worker network interception

rwoll opened the pull request:
Resolves #1090.

Some known TODOs:
- [ ] Code Organization
   - [x] move SW code into own CRServiceWorker 
   - [ ] clean up import cycles (and general haphazard importing)
   - [ ] Intro NetworkedTarget to try to mitigate all the `this._page || this._serviceWorker` bits
- [ ] More Tests!
   - [ ] CORS
   - [ ] Native Service Worker Fetch
   - [ ] Service Worker Cache Intercaction
   - [ ] Multi-Page Scenarios
   - [ ] WebSockets
   - [ ] setOffline
   - [ ] multiple
- [x] Toggle Service Worker Interception based on current routeing
- [ ] Review/resolve all `FIXME(raw):`/`TODO(raw):`
- [ ] (maybe) Add option to opt-out of SW interception. The current patch may break users who are currently doing interception and don't expect the SW bits to show up (or they assume `request.frame()` is always valid…it was before!)
- [x] ~(maybe) Add `worker.on('request',...)` etc. events to mirror page events~

rwoll committed:
wip: feat: service worker interception

This is the beginning of fixing
https://github.com/microsoft/playwright/issues/1090.

There's still quite a bit to be done, but this at least shows an
end-to-end path to accomplish the feature.

rwoll committed:
unfocus tests

rwoll committed:
remove assumption that Request et. al. belong to frame

rwoll committed:
fail tests best on browser

rwoll committed:
nit: avoid conditionals in the Request constructor bits

rwoll committed:
test: add stub tests

rwoll commented on tests/browsercontext-route.spec.ts:
```diff
@@ -235,3 +235,122 @@ it('should overwrite post body with empty string', async ({ context, server, pag
   const body = (await req.postBody).toString();
   expect(body).toBe('');
 });
+
```
NB: I'll re-organize the tests at the end before flipping to from draft. They'll likely end up their own file or next to the existing tests they relate (e.g. HAR).

Mone71 commented:
Thanks

dgozman commented on packages/playwright-core/src/client/network.ts:
```diff
@@ -225,7 +228,14 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
-    const page = this.request().frame()._page;
+    // FIXME(raw): SW requests do not have an associated frame: we probably want to signal and check it is explicilty SW request
+    let page = null;
+    try {
```
`const page = this.request()._frame?._page;`

dgozman commented on packages/playwright-core/src/server/frames.ts:
```diff
@@ -268,7 +268,7 @@ export class FrameManager {
 
   requestStarted(request: network.Request, route?: network.RouteDelegate) {
-    const frame = request.frame();
+    const frame = request.frame()!;
```
I'd make this method call page.requestStarted that calls context.requestStarted so that we get a nice chain of methods that always bubble up.

dgozman replied:
I think this might still be relevant.

dgozman replied:
Plan of action:
- inline `page._requestStarted` into `frameManager._onRequestStarted`;
- remove `browserContext._requestStarted` and inline it into `CRServiceWorker._requestStarted`.
We end up with two separate implementations.

rwoll replied:
This should now be fixed.

dgozman commented on packages/playwright-core/src/server/network.ts:
```diff
@@ -97,15 +98,17 @@ export class Request extends SdkObject {
+  constructor(context: contexts.BrowserContext, frame: frames.Frame | null, redirectedFrom: Request | null, documentId: string | undefined,
     url: string, resourceType: string, method: string, postData: Buffer | null, headers: types.HeadersArray) {
-    super(frame, 'request');
+    super(context, 'request');
```
This should be `super(frame || context, 'request')` to preserve attribution to the page.

dgozman commented on tests/browsercontext-route.spec.ts:
```diff
@@ -235,3 +235,122 @@ it('should overwrite post body with empty string', async ({ context, server, pag
+
+it.describe('service workers', () => {
+  it('intercepts initial script', async ({ server, page, context, browserName }) => {
+    it.fail(browserName === 'firefox');
```
This works in webkit? Very surprising!

rwoll replied:
Yes! WebKit this surprised me too! https://github.com/microsoft/playwright/runs/5453708442?check_suite_focus=true

dgozman commented on packages/playwright-core/src/server/chromium/crBrowser.ts:
```diff
@@ -312,9 +313,28 @@ class CRServiceWorker extends Worker {
       this._createExecutionContext(new CRExecutionContext(session, event.context));
     });
+
+    // FIXME(raw): This will all be moved into a crServiceWorkerNetworkManager that will use shared bits that page/frame-level network mgr. (currently crNetworkManager) will depend on
```
I'd rather reuse CRNetworkManager as is. This way we'll get all the nice logic in there, and will also see where else we should hook up service workers (e.g. `authenticate` or `setOffline`).  What are the blockers for that?

rwoll replied:
The big one is ServiceWorker's won't be intrinsically tied to a `page` (or `frame`) and CRNetworkManager is frame/page-centric as is. The alternative motive was to see which bits (and nuance) of CRNetworkManager are absolutely necessary in the context of SW. Building it from first principles also gives me an opportunity to better understand how/when the nuanced bits of CRNetworkManager become necessary (so in this way it's more of a learning exercise). Although, in the end I may end up throwing away CRServiceWorkerNetworkManager and just taking the learning along! As I get a bit deeper, I may cut my learning exercise short since the convergence looks more obvious than it did before I started! 

@pavelfeldman and I discussed this topic a bit so CC'ing in case he has comments to add!

dgozman reviewed (commented):
Looks nice overall!

rwoll commented on packages/playwright-core/src/server/firefox/ffNetworkManager.ts:
```diff
@@ -149,7 +149,7 @@ export class FFNetworkManager {
 }
 
-const causeToResourceType: {[key: string]: string} = {
+const causeToResourceType: { [key: string]: string } = {
```
Oops! Will remove these from final merge. I had an errant auto-format enabled on my editor, and it coincidentally differs but does not violate eslint rules.

mxschmitt replied:
If you find a eslint rule for that which aligns it, we're more than happy to add it. Because currently there is none configured and vscode formatting != whats used in the code base and eslint accepts both.

rwoll committed:
test: add more placeholder tests

rwoll committed:
tighten SDKObject scopes

Meir017 commented:
adding a method on the service-worker's request/response that references the page/worker request would be helpful as well here.
puppeteer has a boolean flag for this https://pptr.dev/#?product=Puppeteer&version=v9.0.0&show=api-httpresponsefromserviceworker, playwright does not have any indicator about service worker on the request/response object, this should not be a breaking change

as can be seen in the devtools here https://googlechrome.github.io/samples/service-worker/basic/
![image](https://user-images.githubusercontent.com/9786571/157170947-462784ac-82bc-46fa-94e5-f43adfbf3c2b.png)

![image](https://user-images.githubusercontent.com/9786571/157171042-750faf17-85c1-4811-b87d-73fc2cc28d0b.png)
![image](https://user-images.githubusercontent.com/9786571/157171055-fd6bc84f-d421-4c83-a43f-9bbff0a14c68.png)

Doug-Bowen commented:
Is there a general ETA on this work.  This is crucial to an issue im solving now.  If this is slated for a later release, I'll move in a different direction for the time being.  

No pressure, just curious.

rwoll commented:
> Is there a general ETA on this work. This is crucial to an issue im solving now. If this is slated for a later release, I'll move in a different direction for the time being.
> 
> No pressure, just curious.

@Doug-Bowen Thanks for your interest! I can't commit to an ETA. Please subscribe to #1090 to be notified of when it lands. Once it's in the main branch, you'll be able to try it out via `@next`.

rwoll committed:
wip: use CRNetworkManager for Service Workers, too

There's still lots that needs to be changed and cleaned up. Initial
attempts tried to get too much working at once (and create pretty but
possibly poor abstractions), so this backs off the changes to a minimal
set.

NB: This patchset requires a CR >= r979924 (https://chromium-review.googlesource.com/c/chromium/src/+/3510917).

To use a local copy, be sure to set `CRPATH` like so:

```
$ CRPATH=$HOME/code/google/chromium/src/out/Default/Chromium.app/Contents/MacOS/Chromium npm run ctest -- --reporter=line
```

rwoll commented on packages/playwright-core/src/server/chromium/crPage.ts:
```diff
@@ -692,6 +692,13 @@ class FrameSession {
+    if (event.targetInfo.type === 'service_worker') {
+      // const work = this._networkManager
+      const worker = this._crPage._browserContext._browser._serviceWorkers.get(event.targetInfo.targetId)!;
+      assert(worker, `Expected to find a worker for ${event.targetInfo.url}`);
```
Like this whole MR, there's a lot of unknowns! This assertion is failing a bunch in our trace-viewer tests, so will need to dig in on what assumptions we can/can't make here about timing/ordering of events.

rwoll committed:
wip: experiment with attachment fix

rwoll committed:
allow init to fail (SW might detach before we get to it)

rwoll committed:
wip: add initial HAR bits, cleanups, tests

rwoll committed:
test: service worker interception with importScripts

rwoll committed:
move CRServiceWorker to own file

rwoll committed:
dynamically set interception

rwoll committed:
nit: listener registration

dgozman commented on tests/har.spec.ts:
```diff
@@ -23,7 +23,7 @@ import type { BrowserContext, BrowserContextOptions } from 'playwright-core';
 import type { Log } from '../packages/playwright-core/src/server/supplements/har/har';
 
-async function pageWithHar(contextFactory: (options?: BrowserContextOptions) => Promise<BrowserContext>, testInfo: any, outputPath: string = 'test.har') {
+export async function pageWithHar(contextFactory: (options?: BrowserContextOptions) => Promise<BrowserContext>, testInfo: any, outputPath: string = 'test.har') {
```
If we export this, let's make it a fixture? Importing specs from other specs is error-prone.

rwoll replied:
Sounds good! Re-located as a fixture. Depending on what tests I end up with, I might put the SW HAR test alongside the other HAR bits. (During dev, it's just nice to have all them in one file 😄 )

rwoll committed:
add SW barrier, fill in more of HAR

rwoll committed:
nit: use alias for TODOs

mxschmitt commented on tests/chromium/chromium.spec.ts:
```diff
@@ -67,6 +67,34 @@ test('should intercept service worker requests (main and within)', async ({ cont
 });
 
+test('should intercept service worker importScripts', async ({ context, page, server }) => {
+  context.route('**/import.js', route =>
```
route is async (the other places too)

rwoll committed:
fix leaky SW listener

rwoll committed:
address review feedback

rwoll commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -472,7 +500,7 @@ class InterceptableRequest {
 }
 
-class RouteImpl implements network.RouteDelegate {
+export class RouteImpl implements network.RouteDelegate {
```
Nope! Thanks!

rwoll commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -58,13 +70,14 @@ export class CRNetworkManager {
-      eventsHelper.addEventListener(session, 'Network.webSocketFrameReceived', e => e.response.payloadData && this._page._frameManager.webSocketFrameReceived(e.requestId, e.response.opcode, e.response.payloadData)),
-      eventsHelper.addEventListener(session, 'Network.webSocketClosed', e => this._page._frameManager.webSocketClosed(e.requestId)),
-      eventsHelper.addEventListener(session, 'Network.webSocketFrameError', e => this._page._frameManager.webSocketError(e.requestId, e.errorMessage)),
+      // TODO(raw): Instrument WebSocket for Service Workers.
```
Updated! `raw` are my initials, but I'll use my GH handle instead for consistency!

rwoll committed:
tests: properly await routing

dgozman commented on packages/playwright-core/src/protocol/protocol.yml:
```diff
@@ -2356,7 +2356,8 @@ Request:
   initializer:
-    frame: Frame
+    frame: Frame?
+    serviceWorker: Worker?
```
Remove this field, it's unused.

dgozman commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -58,14 +68,21 @@ export class CRNetworkManager {
+
+    if (this._page) {
+      listeners.push(...[
+        eventsHelper.addEventListener(session, 'Network.webSocketCreated', e => this._page?._frameManager.onWebSocketCreated(e.requestId, e.url)),
```
```js
const page = this._page;
...e => page._frameManager....
```

dgozman commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -158,10 +175,23 @@ export class CRNetworkManager {
     if (!event.responseStatusCode && !event.responseErrorReason) {
-      // Request intercepted, deliver signal to the tracker.
-      const request = this._requestIdToRequest.get(event.networkId!);
+      const request = this._requestIdToRequest.get(event.networkId!) || this._maybeAdoptServiceWorkerMainRequest(event.networkId!);
```
Add a test that `'request'` event is dispatched for the main sw script without interception, and is not dispatched on the page.

dgozman commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -40,15 +43,22 @@ export class CRNetworkManager {
+  setParentManager(parent: CRNetworkManager) {
+    if (this._parentManager && this._parentManager !== parent)
+      throw new Error('A different parent manager was already set. Cannot overwrite.');
+    this._parentManager = parent;
```
This seems like a nice place to call `_maybeAdoptServiceWorkerMainRequest` and call `_onRequestWillBeSent` with that event.

dgozman commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -158,10 +175,23 @@ export class CRNetworkManager {
+  // We assume these came in the the same session (on the Service Worker), so shuffle things around.
+  // In the future, this may be resolved; see https://crbug.com/1304536
+  _maybeAdoptServiceWorkerMainRequest(id: string) {
+    const event = this._parentManager?._requestIdToRequestWillBeSentEvent.get(id);
```
Let's check the `'service-worker'` header and ignore this request inside the page's network manager.

dgozman commented on packages/playwright-core/src/server/chromium/crPage.ts:
```diff
@@ -723,6 +727,23 @@ class FrameSession {
     this._networkManager.instrumentNetworkEvents(session, this._page._frameManager.frame(this._targetId)!);
   }
 
+  private _setParentNetworkManagerForServiceWorker(event: Protocol.Target.attachedToTargetPayload) {
```
Let's move this to BrowserContext.

dgozman commented on packages/playwright-core/src/server/supplements/har/har.ts:
```diff
@@ -56,7 +56,8 @@ export type Entry = {
   connection?: string;
   _requestref: string;
-  _frameref: string;
+  // FIXME(rwoll): Think through this as it might be a breaking change; if so, use '' for Service Workers
```
That's fine, let's drop the FIXME.

dgozman commented on packages/playwright-core/src/server/supplements/har/harTracer.ts:
```diff
@@ -128,11 +128,14 @@ export class HarTracer {
   }
 
-  private _addBarrier(page: Page, promise: Promise<void>) {
+  private _addBarrier(target: { on: (event: 'close', fn: () => void) => void } | null, promise: Promise<void>) {
```
`target: Page | Worker`

dgozman commented on packages/playwright-core/src/server/supplements/har/harTracer.ts:
```diff
@@ -404,7 +410,8 @@ export class HarTracer {
 }
 
-function createHarEntry(method: string, url: URL, requestref: string, frameref: string): har.Entry {
+// TODO(rwoll): Consider adding _serviceworkerref
```
Let's not do that! (c) Ross

rwoll committed:
review feedback; mechanical cleanups

rwoll committed:
move _setParentNetworkManagerForServiceWorker onto CRBrowserContext

rwoll committed:
inline _requestStarted

rwoll commented:
Addressed the mechanical fixes/improvements. Will address the `_maybeAdoptServiceWorkerMainRequest` bits of feedback tomorrow!

This is getting near feature complete for an initial land modulo some additional testing. 😄

rwoll committed:
ensure only Service Worker-identified requests are adoptable

Meir017 commented on tests/chromium/chromium.spec.ts:
```diff
@@ -30,6 +30,118 @@ test('should create a worker from a service worker', async ({ page, server }) =>
   expect(await worker.evaluate(() => self.toString())).toBe('[object ServiceWorkerGlobalScope]');
 });
 
+test('should create a worker from service worker with noop routing', async ({ context, page, server }) => {
```
I think we should also verify that `page` does not report the requests sent from within the service worker, only the request being sent from the page (being intercepted by the service-worker)

rwoll replied:
> I think we should also verify that page does not report the requests sent from within the service worker, only the request being sent from the page (being intercepted by the service-worker)

@Meir017 Let me make sure I'm following!

So if `example.html` looks like:

```html
<script>
  navigator.serviceWorker.register('/worker.js');
  navigator.serviceWorker.ready.then(async registration => {
      await fetch('/payload.txt');
  });
</script>
```

and `worker.js` looks like:

```js
fetch('/top-level-sw-request.txt');
self.addEventListener('fetch', event => {
  if (!event.request.url.endsWith('/payload.txt')) return;
  event.respondWith(fetch(event.request));
});
```

On which Playwright object (`page` or `serviceWorker`) would you expect to see the `request` event fire for each of the two requests (`/payload.txt` and `/top-level-sw-request.txt`)?

Meir017 replied:
I would expect to see two `request` events fired. one for the page and another for the service worker.
this can be seen in the network tab of the browser.

from your example
- page request `/worker.js`
- sw request `/top-level-sw-request.txt`
- page request `/payload.txt`
- sw request `/payload.txt`

see my comment https://github.com/microsoft/playwright/pull/12565#issuecomment-1061419170

rwoll committed:
simplify adoptions (assuming upstream bug gets fixed)

rwoll commented on packages/playwright-core/src/server/chromium/crNetworkManager.ts:
```diff
@@ -118,6 +136,19 @@ export class CRNetworkManager {
+        // Service Worker Network.requestWillBeSent event fires on the parent frame session,
+        // but the Fetch.requestPaused comes in on the Service Worker session.
+        // We assume these came in the the same session (on the Service Worker), so shuffle things around.
+        // In the future, this may be resolved; see https://crbug.com/1304536
```
Currently fixing this upstream.

rwoll committed:
remove adoption code (depends on upstream CR patch)

Depends on https://chromium-review.googlesource.com/c/chromium/src/+/3526571

Set CRPATH in .env before running.

rwoll commented:
Closing this! Will re-open once CR upstream changes are all landed and this code is reworked a bit around some internal refactors and conflicts.
